File: solutions/841. Keys and Rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        set<int> visitedRooms ;
        visitedRooms.insert(0);
​
        stack<int> cacheRooms ;
        cacheRooms.push(0);
        while (cacheRooms.empty()==false) {
            vector<int> roomKeys = rooms[cacheRooms.top()];
            cacheRooms.pop();
            for (int key : roomKeys) {
                if (visitedRooms.count(key)==0) {
                    visitedRooms.insert(key);
                    cacheRooms.push(key);
                }
            }
        }
        return (int) visitedRooms.size() == (int) rooms.size() ;   
    }
};

File: solutions/841. Keys and Rooms.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        int totalRooms = rooms.size();
        boolean[] visitedRooms = new boolean[totalRooms];
        exploreAllRoomsAndOpen(rooms, visitedRooms, 0);
        return hasAllExplored(visitedRooms);
    }
​
    private boolean hasAllExplored(boolean[] visitedRooms) {
        boolean status = visitedRooms[0];
        for (boolean val : visitedRooms)
            status = status & val;
        return status;
    }
​
    private void exploreAllRoomsAndOpen(List<List<Integer>> rooms, boolean[] visitedRooms, int currentRoom) {
        visitedRooms[currentRoom] = true;
        for (int nextRoomKey : rooms.get(currentRoom)) {
            if (!visitedRooms[nextRoomKey]) {
                exploreAllRoomsAndOpen(rooms, visitedRooms, nextRoomKey);
            }
        }
    }
}
